### ISSUE TITLE
OMEMO messages directly to self generate an erroneous error

### ISSUE THREAD

**Issue by GigabyteProductions** [gkdr/lurch#99]
As of 530c61474ba0b9e5542e2bf936f75988df93da15 and gkdr/carbons@153b45c425894d1990a2fb7237fdc0309023cacf. Only applies if carbons is enabled. Only applies if message is OMEMO-encrypted.

Two devices logged into my account, messaging my _own_ JID.

Chat window (notice the message is, in fact, successfully decrypted):
```
(07:19:02 PM) gigabyteproductions@gigabyteproductions.net/8229338302065577985163090: test 4
(07:19:02 PM) There was an error decrypting an OMEMO message addressed to this device. See the debug log for details.
```

Debug window:
```
(19:19:02) jabber: Recv (ssl)(1064): <message to='gigabyteproductions@gigabyteproductions.net/8229338302065577985163090' from='gigabyteproductions@gigabyteproductions.net' type='chat'><sent xmlns='urn:xmpp:carbons:2'><forwarded xmlns='urn:xmpp:forward:0'><message xml:lang='en' to='gigabyteproductions@gigabyteproductions.net/phone' from='gigabyteproductions@gigabyteproductions.net/2114696233876654593163138' type='chat' id='purplea5eb9619' xmlns='jabber:client'><encrypted xmlns='eu.siacs.conversations.axolotl'><header sid='1195612485'><key rid='967469621'>MwohBTiKTtkhEdRhYYXIlBre23TSCSz5wfg4pp7Mur5lbzkWEAEYBCIw21tJ2Lode1AYE5ZT+LdXGqLku3KMw7vdRjSQcGDLfU087o/u7uorMgkoF8h6zluB5BEr3cAY7no=</key><key rid='1615521877'>MwohBX2mSL50OD3pTeXuL7Qj86vD9H+QDQnjhfMaew3qfIpGEAEYACIwn69oaceai73d+xhTVzfHtatI9rvkOjqA+O1MXvvNoxPPmnLhfOZyCLzrr4/HtHzn1WHYjRksnTo=</key><iv>ufINRzUAN3wA+GMH/OgEkw==</iv></header><payload>X2J8I01+</payload></encrypted><encryption xmlns='urn:xmpp:eme:0' namespace='eu.siacs.conversations.axolotl' name='OMEMO'/><store xmlns='urn:xmpp:hints'/></message></forwarded></sent></message>
(19:19:02) jabber: XML parser error for JabberStream 0x5650a325c950: Domain 3, code 100, level 1: xmlns: URI eu.siacs.conversations.axolotl is not absolute
(19:19:02) carbons: Received carbon copy of a sent message.
(19:19:02) carbons: Carbon copy of sent message does not contain a body - stripping and passing it through.
(19:19:02) carbons: Carbon copy of sent message contains a body, but also an additional encrypted element - stripping and passing it through.
(19:19:02) jabber: Recv (ssl)(1072): <message to='gigabyteproductions@gigabyteproductions.net/8229338302065577985163090' from='gigabyteproductions@gigabyteproductions.net' type='chat'><received xmlns='urn:xmpp:carbons:2'><forwarded xmlns='urn:xmpp:forward:0'><message xml:lang='en' to='gigabyteproductions@gigabyteproductions.net/phone' from='gigabyteproductions@gigabyteproductions.net/2114696233876654593163138' type='chat' id='purplea5eb9619' xmlns='jabber:client'><encrypted xmlns='eu.siacs.conversations.axolotl'><header sid='1195612485'><key rid='967469621'>MwohBTiKTtkhEdRhYYXIlBre23TSCSz5wfg4pp7Mur5lbzkWEAEYBCIw21tJ2Lode1AYE5ZT+LdXGqLku3KMw7vdRjSQcGDLfU087o/u7uorMgkoF8h6zluB5BEr3cAY7no=</key><key rid='1615521877'>MwohBX2mSL50OD3pTeXuL7Qj86vD9H+QDQnjhfMaew3qfIpGEAEYACIwn69oaceai73d+xhTVzfHtatI9rvkOjqA+O1MXvvNoxPPmnLhfOZyCLzrr4/HtHzn1WHYjRksnTo=</key><iv>ufINRzUAN3wA+GMH/OgEkw==</iv></header><payload>X2J8I01+</payload></encrypted><encryption xmlns='urn:xmpp:eme:0' namespace='eu.siacs.conversations.axolotl' name='OMEMO'/><store xmlns='urn:xmpp:hints'/></message></forwarded></received></message>
(19:19:02) jabber: XML parser error for JabberStream 0x5650a325c950: Domain 3, code 100, level 1: xmlns: URI eu.siacs.conversations.axolotl is not absolute
(19:19:02) carbons: Received carbon copy of a received message.
(19:19:02) lurch: lurch_message_decrypt: failed to decrypt key (-1001)
```

I would guess this is not a fault of carbons, but lurch misinterpreting the stanzas presented by carbons.

**Comment by gkdr** [comment]
This didn't pass through carbons before so it wasn't apparent that the handling is wrong. I'll have a look.

**Comment by gkdr** [comment]
I was able to reproduce it but I'm a bit too tired to think about it properly now.
This works expected - thanks to carbons, you receive the message two times: one sent, one received. Then, since the double ratchet is forward secure, you cannot decrypt the key a second time, as the session has already been moved on and the message key deleted.
I'll have to think about where to add an exception so the decryption failure is ignored.

By the way, what server are you using? I'm running Prosody 0.10 and instead of the "received" carbon copy I receive the original message, which you don't seem to receive.

**Comment by GigabyteProductions** [comment]
I use ejabberd 17.09 from Fedora 27.

**Comment by gkdr** [comment]
`-1001` is an error code from libsignal and I realized I didn't even look up what exactly it means. Luckily it is very specific: `SG_ERR_DUPLICATE_MESSAGE`, i.e. it doesn't even try to decrypt it because it can tell from the counter it's a duplicate. So I guess if decrypting results in this error AND the sender is yourself, an exception which still logs something just in case isn't that misplaced.

A funny thing happened, though: I cannot reproduce this any longer. I don't think I updated Prosody (but I might have), but carbon-copies of sent messages simply don't have a recipient anymore so I can't read any messages directed to myself...

**Comment by GigabyteProductions** [comment]
Would you like a test account on my server?

**Comment by gkdr** [comment]
It's fine, I just needed to complain somewhere :D It seems I "broke" it myself in 2ea47f488ba80910121f4f007f6c56400e856509.
If you want, you can try 97b4c1566958745ef0d21a02351720aa1b27524d and see if it fixes the issue for you.

**Comment by GigabyteProductions** [comment]
Works without chat error!
```
(11:06:35) lurch: ignoring decryption error due to a duplicate message from own account to own account
```

**Comment by gkdr** [comment]
This fix was merged in 0.6.8, but it seems I forgot to close the issue, which I am doing now.
